### PR TITLE
Core/Players: re-organize SetSkill method to handle initialized skills correctly

### LINKED ISSUE #24107: Core/Player: Profession skill not saved in database (Bounty)
**Description:**
if i learn the profession that work but i can progress the skill but if reconnect to the server that not saved in the database.

**Steps to reproduce the problem:**

1. create any characters  
2. learn profession tailoring
3. progress this profession
4. ex. 39/75 (skill)
5. disconnect the character and connect again.
6. the progression restart to 1/75 skill.

**Branch(es):**
Master, Last updated

**Operating system:** WinServer2016 x64

### PR DESCRIPTION
**Pull request by Ovahlord**
**Changes proposed:**
-  updated the logics in SetSkill to support the new skillmap handlings correctly

**Target branch(es):** 3.3.5/master
- [ ] 3.3.5
- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)
closes #24107

**Tests performed:** (Does it build, tested in-game, etc.)
tested on my 434 fork

### CODE REVIEW

**Comment by buddyqc69** [comment]
i can learn all skill of tailoring one shot ?? lol

**Comment by buddyqc69** [comment]
ex. i learn the tailoring, yeah but 0/0 for that skill, but if i learn with .cast or .learn 264617 the skill work for save with progress level of the skill. but im expecting an trouble with de database or the code, i need to reconnect for change to the next book skill.

**Comment by buddyqc69** [comment]
If learn 264617 is ok that save but if reach up to 50/300 i need relog for update my skill

**Comment by finestlyker** [comment]
[a14b75fd0202_worldserver.exe_[29-1_21-12-15].txt](https://github.com/TrinityCore/TrinityCore/files/4131846/a14b75fd0202_worldserver.exe_.29-1_21-12-15.txt)

i have try you fix

if i want to create an DK, that crash the server
if i want to create tauren, the tauren is naked, lol

**Comment by buddyqc69** [comment]
@finestlyker what the version you use ?

**Comment by Ovahlord** [comment]
Your crashlog is useless compile in debug and try again.

**Comment by buddyqc69** [comment]
I have the same problem like @finestlyker

**Comment by Aokromes** [comment]
> I have the same problem like @finestlyker

then supply a debug mode crashlog.

**Comment by buddyqc69** [comment]
i have the same trouble, after create DK i have this warning in the console

> Tried to add skill 95 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 183 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 162 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 109 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 160 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 229 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 413 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 414 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 415 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 44 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 756 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 43 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 137 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 54 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 55 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 293 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 118 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 762 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 777 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 172 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 796 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 810 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 934 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills
> Tried to add skill 2727 but player Testrace (GUID Full: 0x0800040000000000000000000000001C Type: Player Entry: 0 Low: 28) cannot have additional skills

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -5638,25 +5648,24 @@ void Player::SetSkill(uint16 id, uint16 step, uint16 newVal, uint16 maxVal)
...
+                break;
             }
         }
 
         if (!skillSlot)
         {
-            TC_LOG_ERROR("misc", "Tried to add skill #%u but the player cannot have additional skills", id);
+            TC_LOG_ERROR("misc", "Tried to add skill %u but player %s (%s) cannot have additional skills", id, GetName().c_str(), GetGUID().ToString().c_str());
             return;
         }
```
Comment (buddyqc69) [inline src/server/game/Entities/Player/Player.cpp]: we have remove this code and the server doesnt not crash after

**Review by buddyqc69 — COMMENTED** [review]
PLAYER_MAX_SKILLS because is not undefined ?

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (Ovahlord) [reply src/server/game/Entities/Player/Player.cpp]: Dude, would you kindly give me a proper crashlog already?

Comment (buddyqc69) [reply src/server/game/Entities/Player/Player.cpp]: it does not want to work currently it says that there are missing files in debug mode

Comment (Faq) [reply src/server/game/Entities/Player/Player.cpp]: so basically @buddyqc69 meant he cant compile core in debug to get proper crashlog.

Comment (Ovahlord) [reply src/server/game/Entities/Player/Player.cpp]: Somehow I have a feeling that he isn't even using the official TC branch. I would like to get a test result on a actual TC version so I can merge this thing at last.

Comment (Lillecarl) [reply src/server/game/Entities/Player/Player.cpp]: We should do some cmake magic to disable release build targets :p

Comment (buddyqc69) [reply src/server/game/Entities/Player/Player.cpp]: nope i use the latest trinitycore :)

**Comment by Ovahlord** [comment]
issues are solved. tested it myself this time. all is working fine now.

**Pull request merged**